### ISSUE TITLE
Wrong/unclear exp and sqrt learning_rate_decay_scheme

### ISSUE THREAD

**Issue by martinpopel** [tensorflow/tensor2tensor#517]
See utils/optimize.py:[learning_rate_decay](https://github.com/tensorflow/tensor2tensor/blob/a1d7ed7e7d96ea0d89ca153ba46b7f51d394580f/tensor2tensor/utils/optimize.py#L134-L193)

* It is called only from one place ([t2t_model.py](https://github.com/tensorflow/tensor2tensor/blob/a1d7ed7e7d96ea0d89ca153ba46b7f51d394580f/tensor2tensor/utils/t2t_model.py#L289)), so the parameters `num_worker_replicas=1` and `num_train_steps=1` have always their default values and are thus misleading.
* Even if it works, multiplying warmup_steps by num_worker_replicas (asynchronous only?) under the hood may be a bit unexpected. I think it should be done more visibly, e.g. in the hyperparameters, so when I want to override the warmup_steps to N, I can do that directly without having to counteract the hidden magic and set counter-intuitive `N / num_worker_replicas`.
* `exp100k` ignores the warmup steps unlike `exp10k`, `exp50k` and `exp500k`, which is quite misleading.
* `_exp_decay_after` takes parameters `step` and `from_which_step`, but depends actually only on their difference. So it is misleading to subtract `warmup_steps` from *both* parameters when calling `_exp_decay_after` - it just obfuscates the code without having any effect.
* `exp10k`, `exp50k` and `exp500k` do not work as expected (they result in almost zero learning rate) because e.g. in case of `exp10k` the main formula `rate**(step - from_which_step)` boils down to  `0.9995**(step + 9999)` because `from_which_step` includes `num_train_steps - 10000`, while I think it should be just 10000.
* The `noam` scheme has maximum learning rate about 1.7 (after warmup_steps=8k, the exact values slightly differ based on hidden_size and warmup_steps) and obviously it works well (as `noam` is the default in most problems). However, the `sqrt` scheme has maximum learning_rate=500 (after warmup_steps) because there is an artificial multiplier 500. I think the person adding the multiplier did not expect it will be called as `_sqrt_decay(step - warmup_steps)`. If it was `_sqrt_decay(step)` than with warmup_steps=8k, the maximum would be about 5.5 and with warmup_steps=16k about 4, which sounds more reasonable.

I suggest to delete or fix all the schemes which are broken now, so they don't mislead the users.

**Comment by rsepassi** [comment]
Thank you for bringing this up. I agree that these schemes are really weird right now and we should probably delete many of them and clean up and correct the remaining ones.

**Comment by martinpopel** [comment]
Some of the problems are resolved in 1.4.3, some are still there and several new problems were introduced (#555).

**Comment by rsepassi** [comment]
How are things in 1.4.4? I believe everything is ok now.

**Comment by martinpopel** [comment]
In fact, most of the problems are still there in 1.4.4 (just from seeing the source code without actually running it) and I found few more potential problems. Here is the updated list:

- [ ] [learning_rate_schedule](https://github.com/tensorflow/tensor2tensor/blob/80b2f7300ce2021b9ca0a7877b2670bbdea2b2bf/tensor2tensor/utils/optimize.py#L223) is still called only from one place (`t2t_model.py`), so the parameter `num_worker_replicas` always has the default value of 1 and I still consider this misleading.
- [ ] `num_worker_replicas` is used only to divide the `warmup_steps`, which is probably meant only for the asynchronous distributed training (while the default is the synchronous training). I still consider this misleading because then warmup steps do not correspond to what one sees in TensorBoard etc.
- [ ] The previous point opens a more interesting question (which perhaps should be discussed in a different issue): should the length of the warmup phase depend on the number of steps or on the number of examples seen? Maybe we should use a parameter called `learning_rate_warmup_examples` instead of `learning_rate_warmup_steps`. Similarly with the `learning_rate_decay` (exp, sqrt,...) - now it depends on the steps, but in most papers it depends on the number of epochs. I think we should do `examples = global_step * batch_size * number_of_GPUs / constant` and then use `examples` instead of `global_step` within the `learning_rate_decay` function.
- [ ] The `_exp_decay_after` function is not used anymore and can be deleted.
- [ ] The `sqrt` scheme has maximum learning_rate=500 (after warmup_steps) because there is an artificial multiplier 500. I think the person adding the multiplier did not expect it will be called as `_sqrt_decay(step - warmup_steps)`. If it was `_sqrt_decay(step)` than with `warmup_steps=8k`, the maximum would be about 5.5 and with warmup_steps=16k about 4, which sounds more reasonable.
- [ ] The `linear_warmup_rsqrt_decay` aka `noam` scheme is not parametrized in a user-friendly way. When lowering the `warmup_steps` and keeping the `learning_rate`, I would expect that the maximal learning rate remains the same (it will just occur sooner), but actually the maximal learning rate is increased. To counter-act this, one has to lower the `learning_rate` with a complicated formula. My suggestion is to change the current behavior (let's keep the backward compatibility aside for now) and interpret the `learning_rate` parameter as the maximum learning rate in all schedules (that is make sure the maximum is always 1 within the `learning_rate_schedule` function).

**Comment by lukaszkaiser** [comment]
@martinpopel : I think you're right in all of the points. We'll get to them slowly, but don't hesitate to send PRs in the meantime if you have code that repairs it. Thanks!
